extensions/amp-autocomplete/0.1/test/test-amp-autocomplete.js

```diff
@@ -245,6 +247,46 @@ describes.realWin('amp-autocomplete unit tests', {
         'Unexpected filter: invalid');
   });
 
+  it('tokenizeString_ should return an array of tokens', () => {
+    expect(impl.tokenizeString_('')).to.have.ordered.members(['']);
+    expect(impl.tokenizeString_('a b c')).to.have.ordered.members(
+        ['a', 'b', 'c']);
+    expect(impl.tokenizeString_('a-b-c')).to.have.ordered.members(
+        ['a', 'b', 'c']);
+    expect(impl.tokenizeString_('a. ...b).c')).to.have.ordered.members(
+        ['a', 'b', 'c']);
+  });
+
+  it('mapFromTokensArray_ should return a map of token counts', () => {
+    expect(impl.mapFromTokensArray_([])).to.be.empty;
+    expect(impl.mapFromTokensArray_(['a', 'b', 'c'])).to.have.all.keys(
+        'a', 'b', 'c');
+    expect(impl.mapFromTokensArray_(['a', 'b', 'c', 'a', 'a']))
+        .to.have.all.keys('a', 'b', 'c');
+  });
+
+  it('tokenPrefixMatch_ should exhaustively match on complex cases', () => {
+    const item = 'washington, district of columbia (d.c.)';
+    expect(impl.tokenPrefixMatch_(item, 'washington')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'district of columbia')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'of colum')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item,
+        'district washington columbia of')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'dc')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'washington dc')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'washington, d.c.')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'washington, (dc)')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'w.a.s.h.i.n.g.t.o.n.')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'district-of-columbia')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, "washington 'dc'")).to.be.true;
+    expect(impl.tokenPrefixMatch_(item,
+        'washin.gton,   (..dc)+++++')).to.be.true;
+    expect(impl.tokenPrefixMatch_(item, 'ashington dc')).to.be.false;
+    expect(impl.tokenPrefixMatch_(item, 'washi distri')).to.be.false;
+    expect(impl.tokenPrefixMatch_(item, 'columbia columbia')).to.be.false;
+    expect(impl.tokenPrefixMatch_(item, 'd c')).to.be.false;
```

Comment:
so thorough! 👏 

@andrewwatterson @nainar do these match your understanding of how token the prefix filter should work?

Reply:
Bammo! I love these examples.

And yes, these all match the algorithm I previously proposed. I feel more strongly about some rather than others, but I don't hate any of them. (In particular, if I were to relax one it'd be matching on multiple token prefixes, so that 285 and 287 were also true, but I feel like @caroqliu may have already come up with a good reason why we can't do that.)

Reply:
285 - I'm not against returning true on this example; it's just a non trivial change that wasn't an immediately clear priority.

287 - This is a consequence of choosing between removing and splitting by `'.'` If the original term includes `"(D.C.)"` it can either be tokenized into `"dc"` or `"d", "c"`. It doesn't make sense to match on the union of these tokens `"dc", "d", "c"` because then it could match on an input like `"dc d"` which doesn't meaningful equate to `"(D.C.)"`. So when choosing between the two, the former just seemed like a more common/likely user input.

Reply:
you made the right call! no change necessary.